department-of-veterans-affairs/vets-website
# [CST] Only show titles for claims once loading completes.

## Summary

This is a follow-up to #31644; after deploying that change (which showed a more detailed title for some dependency claims), issues with the existing loading behavior w/r/t claim titles were much more noticeable. While the titles that appear on the cards on the list view and the main heading on the detail view only render once the claim has loaded, the document (browser tab) title and the breadcrumbs both render while loading and show:
* default text, e.g. `Status of Your Claim` in the document title and `Status of your disability compensation claim` in a breadcrumb
* text related to a previously loaded claim, i.e. when navigating between claims, while the new claim loads, the titles for the previous claim render for the document title and breadcrumbs

Resolving the problem with default text was relatively simple:
- For the document title, we don't make any updates until the claim has loaded.
  - When loading for the first time, it will show the default application title, `Track Claims`
  - When navigating, it will show the previous page title, which is typically the claims list page title, `Check your claim, decision review, or appeal status`
- For the breadcrumbs, instead of showing a placeholder crumb for the claim detail page, we only show the crumbs for its parent, the claims list page. This way the UI appears to append the relevant crumb once the claim has loaded, instead of a flash of incorrect content followed by the correct content.

Resolving the issue with rendering previous claim data during loading required us to clear claim data (basically resetting the claim state) when leaving a claim detail page, so that the above functionality for the document title and breadcrumbs works as expected regardless of whether a user has previously viewed a different claim.

Additionally, we now only show default text when we fail to load a claim:
- For the document title, we render `Status of Your Claim` (or `Files for Your Claim`, etc.).
- For the breadcrumbs, we render `Status of your claim` (or `Files for your claim`, etc..

## Related issue(s)

* Original issue: https://github.com/department-of-veterans-affairs/va.gov-team/issues/86080
* Original PR: https://github.com/department-of-veterans-affairs/vets-website/pull/31644

## Testing done

When verifying the bug and the solution locally, you'll want to throttle your network connection speed; otherwise loading the mock data is pretty fast. Alternatively, for the document title, you can inspect the `<title>` element under the `<head`> and setup a break on subtree modifications to step through the loading process.

To reproduce the bug on `main`:
- Navigate from the claims list page to a claim detail page and verify that there's a flash of default content in the breadcrumbs and document title while the claim loads.
- Navigate back to the claims list page and to another claim detail page. Verify that there's a flash of previous claim content for the document title and breadcrumbs while the claim loads.

To verify changes are working on this branch:
- Navigate from the claims list page to a claim detail page and verify that the document title and breadcrumbs render as expected (document title shouldn't change until the claim loads, and breadcrumbs will show the crumbs for the list page and add a crumb for the detail page once loaded).
- Navigate back to the claims list page and to another claim detail page. Verify that the document title and breadcrumbs behave the same way they did on the first claim.

## Screenshots

Screenshotting document titles during loading is rather difficult, but here are screenshots for the breadcrumbs
|         | Before | After |
| ------- | ------ | ----- |
| During initial load | <img width="1101" alt="Screenshot 2024-09-06 at 12 23 18 AM" src="https://github.com/user-attachments/assets/e76c9fb3-0d7c-40fe-805d-788a046328e6"> | <img width="1101" alt="Screenshot 2024-09-06 at 12 26 38 AM" src="https://github.com/user-attachments/assets/ffec3479-3084-4638-b814-bf46f41b374a"> |
| After initial load | <img width="1101" alt="Screenshot 2024-09-06 at 12 20 53 AM" src="https://github.com/user-attachments/assets/83876a26-6e94-4b21-8ef0-4741aa5db021"> | <img width="1101" alt="Screenshot 2024-09-06 at 12 26 42 AM" src="https://github.com/user-attachments/assets/7a65bded-3760-4c36-ba52-2176452b7d44"> |
| During load of another claim | <img width="1101" alt="Screenshot 2024-09-06 at 12 24 24 AM" src="https://github.com/user-attachments/assets/d806d991-e2f6-43cd-91b1-b97c3c071f58"> | <img width="1101" alt="Screenshot 2024-09-06 at 12 26 51 AM" src="https://github.com/user-attachments/assets/d513e7f9-f20a-4770-8401-22082ef61d03"> |
| After load of another claim |<img width="1101" alt="Screenshot 2024-09-06 at 12 24 29 AM" src="https://github.com/user-attachments/assets/7fa9c0b1-cf90-4dd9-9cbc-9f490b471444"> | <img width="1101" alt="Screenshot 2024-09-06 at 12 26 55 AM" src="https://github.com/user-attachments/assets/88d77f08-c090-4549-9fc9-7def5b2f24ef"> |

## What areas of the site does it impact?

Claim Status Tool

## Acceptance criteria

### Quality Assurance & Testing


### Error Handling


## Requested Feedback

My primary concern here is whether anything in the application relied on previous claim data being in state after a user leaves the detail page. I'm still learning about all the parts of the app and have very little insight into the possible side-effects of this change.

Otherwise, see self-review!
